packages/sdk/src/util/userFilter.ts

```diff
@@ -0,0 +1,88 @@
+/*!
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { ActionHandler, Guid, User } from '..';
+
+/** A callback that handles when a user joins or leaves the MRE session */
+export type UserEntryExitCallback = (user: User) => void;
+
+/** An event type for [[UserFilter.shouldForwardUserEvent]] */
+export type UserInteractionType = 'joined' | 'input';
+
+/** A class that has user joined/left callback hooks. Applies to [[Context]], and also [[UserFilter]]s. */
+export interface UserEntryExitPoint {
+	onUserJoined(callback: UserEntryExitCallback): void;
+	onUserLeft(callback: UserEntryExitCallback): void;
+	offUserJoined(callback: UserEntryExitCallback): void;
+	offUserLeft(callback: UserEntryExitCallback): void;
+}
+
+/** Base class for classes that filter out users from MRE awareness */
+export abstract class UserFilter implements UserEntryExitPoint {
+	private joinedCallbacks = new Set<UserEntryExitCallback>();
+	private leftCallbacks = new Set<UserEntryExitCallback>();
+
+	/** The set of IDs of joined users */
+	protected joinedUsers = new Set<Guid>();
+
+	/**
+	 * Set up the user filter
+	 * @param context An MRE.Context object, or another user filter instance
+	 */
+	constructor(protected context: UserEntryExitPoint) {
+		this.context.onUserJoined(u => this.onUpstreamUserJoined(u));
+		this.context.onUserLeft(u => this.onUpstreamUserLeft(u));
+	}
+
+	/** Register a callback that will be called when a user that passes the filter joins the MRE session */
+	public onUserJoined(callback: UserEntryExitCallback) {
+		this.joinedCallbacks.add(callback);
+	}
+
+	/** Deregister an [[onUserJoined]] callback */
+	public offUserJoined(callback: UserEntryExitCallback) {
+		this.joinedCallbacks.delete(callback);
+	}
+
+	/** Register a callback that will be called when a user that passes the filter leaves the MRE session */
+	public onUserLeft(callback: UserEntryExitCallback) {
+		this.leftCallbacks.add(callback);
+	}
+
+	/** Deregister an [[onUserLeft]] callback */
+	public offUserLeft(callback: UserEntryExitCallback) {
+		this.leftCallbacks.delete(callback);
+	}
+
+	/** Process an input event only from users that pass the filter */
+	public filterInput(eventHandler: ActionHandler<any>): ActionHandler<any> {
+		return (user: User, data: any) => {
+			if (this.shouldForwardUserEvent(user, 'input')) {
+				eventHandler(user, data);
+			}
+		};
+	}
```

Comment:
The thing I don't like about this is that it strips away the strongly typed event data this system has been designed to uphold.  Inserting this filter will strip that away from the event handler that the developer is going to supply to this.

Reply:
Is it not possible to make this a generic function itself and thus propagate the ActionDataT that is originally supplied with the action callback registration method?  Such as `public filterInput<ActionDataT = void>(eventHandler: ActionHandler<ActionDataT>): ActionHandler<ActionDataT>`?

---
In reply to: [445818329](https://github.com/microsoft/mixed-reality-extension-sdk/pull/634#discussion_r445818329) [](ancestors = 445818329)

Reply:
I'll give it a shot.

---
In reply to: [445820208](https://github.com/microsoft/mixed-reality-extension-sdk/pull/634#discussion_r445820208) [](ancestors = 445820208,445818329)

Reply:
I think that would be `T = void`?  At least that is how we due it in the action handler itself.

---
In reply to: [445832051](https://github.com/microsoft/mixed-reality-extension-sdk/pull/634#discussion_r445832051) [](ancestors = 445832051,445820208,445818329)

Reply:
Also can you check if the type propagates through to the action handler being supplied to this function?  Just want to see if the type carries through to the developer code.

---
In reply to: [445861996](https://github.com/microsoft/mixed-reality-extension-sdk/pull/634#discussion_r445861996) [](ancestors = 445861996,445832051,445820208,445818329)

Reply:
Good catch with the null vs void, fixed. And I verified that the VS Code Intellisense picked up the type through the filter, though haven't actually run anything.

---
In reply to: [445864069](https://github.com/microsoft/mixed-reality-extension-sdk/pull/634#discussion_r445864069) [](ancestors = 445864069,445861996,445832051,445820208,445818329)

Reply:
Perfect.  Not too concerned about running.  It should be fine.  More was making sure the intellisense and all was preserved for the event data.  Thanks for updating!

---
In reply to: [445872113](https://github.com/microsoft/mixed-reality-extension-sdk/pull/634#discussion_r445872113) [](ancestors = 445872113,445864069,445861996,445832051,445820208,445818329)